File: javascript/src/Tags/ImageTag.ts
import { Parser } from '..';
import { HtmlHelper } from '../HtmlHelper';
import { HtmlNode } from '../Nodes/HtmlNode';
import { INode } from '../Nodes/INode';
import { PlainTextNode } from '../Nodes/PlainTextNode';
import { ParseData } from '../ParseData';
import { State } from '../State';
import { Tag } from './Tag';

export class ImageTag extends Tag {
    constructor() {
        super();
        this.Token = 'img';
        this.Element = 'div';
        this.MainOption = 'url';
        this.Options = ['url'];
        this.IsBlock = true;
    }

    public FormatResult(_parser: Parser, _data: ParseData, state: State, _scope: string, options: Record<string, string>, text: string): INode {
        let url = text;
        if (options['url']) url = options['url'];
        if (!url.match(/^([a-z]{2,10}:\/\/)/i)) url = 'http://' + url;
        url = HtmlHelper.AttributeEncode(url);

        const classes = ['embedded', 'image'];
        if (this.ElementClass != null) classes.push(this.ElementClass);
        let element = this.Element;

        if (!this.IsBlock) {
            element = 'span';
            classes.push('inline');
        } else {
            state.SkipWhitespace();
        }

        const before = `<${element} class="${classes.join(' ')}">` +
            '<span class="caption-panel">' +
            `<img class="caption-body" src="${url}" alt="User posted image" />`;
        const after = `</span></${element}>`;
        const plainsp = element == 'div' ? '\n' : '';
        const ret = new HtmlNode(before, PlainTextNode.Empty(), after);
        ret.PlainBefore = `${plainsp}[User posted image]${plainsp}`;
        ret.IsBlockNode = element == 'div';
        return ret;
    }

    public Validate(options: Record<string, string>, text: string): boolean {
        let url = text;
        if (options['url']) url = options['url'];
        return !url.includes('<script') && url.match(/^([a-z]{2,10}:\/\/)?([^\]"\n ]+?)$/i) != null;
    }
}

File: csharp/LogicAndTrick.WikiCodeParser/Tags/ImageTag.cs
﻿using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using LogicAndTrick.WikiCodeParser.Nodes;

namespace LogicAndTrick.WikiCodeParser.Tags
{
    public class ImageTag : Tag
    {
        public ImageTag()
        {
            Token = "img";
            Element = "div";
            MainOption = "url";
            Options = new[] {"url"};
            IsBlock = true;
        }

        public override INode FormatResult(Parser parser, ParseData data, State state, string scope, Dictionary<string, string> options, string text)
        {
            var url = text;
            if (options.ContainsKey("url")) url = options["url"];
            if (!Regex.IsMatch(url, "^([a-z]{2,10}://)", RegexOptions.IgnoreCase)) url = "http://" + url;
            url = HtmlHelper.AttributeEncode(url);

            var classes = new List<string>{"embedded", "image"};
            if (ElementClass != null) classes.Add(ElementClass);
            var element = Element;

            if (!IsBlock)
            {
                element = "span";
                classes.Add("inline");
            }
            else
            {
                state.SkipWhitespace();
            }

            var before = $"<{element} class=\"{String.Join(" ", classes)}\">" +
                         $"<span class=\"caption-panel\">" +
                         $"<img class=\"caption-body\" src=\"{url}\" alt=\"User posted image\" />";
            var after = $"</span></{element}>";
            var plainsp = element == "div" ? "\n" : "";
            return new HtmlNode(before, PlainTextNode.Empty, after)
            {
                PlainBefore = $"{plainsp}[User posted image]{plainsp}",
                IsBlockNode = element == "div"
            };
        }

        public override bool Validate(Dictionary<string, string> options, string text)
        {
            var url = text;
            if (options.ContainsKey("url")) url = options["url"];
            return !url.Contains("<script") && Regex.IsMatch(url, @"^([a-z]{2,10}://)?([^]""\n ]+?)$", RegexOptions.IgnoreCase);
        }
    }
}